### PR TITLE
add criteria RangeAggregation

### PR DESCRIPTION
**Pull request by iNem0o**
### 1. Why is this change necessary?

Allow to aggregate data on predefined ranges values for more flexibility in the DAL.
It could allow to provide faceted filters by predefined range for example.

Bound are computed in SQL as in the elasticsearch native range aggregation :
   -> "from" will be compared with greater than or equal
   -> "to" will be compared with lower than

 
#### Core
* Add `RangeAggregation` and `RangeResult` 
* Add range aggregation parsing and hydration in `EntityAggregator`
* Add range aggregation handling in AggregationParser
* Add test on the aggregation range using database
* Add test on the range aggregation in the AggregationParser

#### Elasticsearch
* Add range aggregation parsing in `CriteriaParser`
* Add range aggregation hydration in `ElasticsearchEntityAggregatorHydrator`
* Add `ElasticsearchRangeAggregation` to manage the conversion of DAL aggregation to elasticsearch
* Add test on the aggregation range using elasticsearch

### 3. Describe each step to reproduce the issue or behaviour.
```
$criteria = new Criteria();
$criteria->addAggregation(
    new RangeAggregation(
        'test-range-aggregation', 'price.gross',
        [
            ['from' => 5, 'to' => 10],
            ['from' => 10, 'to' => 15],
            ['from' => 15, 'to' => 30]
        ]
    )
);
```

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/workflow/2020-08-03-implement-New-Changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### 6. Additional informations

- ECS has automatically fixed style in the changed files
- I had to fix some existing type casting issues noted by phpstan

### CODE REVIEW

File: changelog/_unreleased/2022-09-02-add-range-aggregation.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+author_github: inem0o
+---
+# Core
+* Add `RangeAggregation` and `RangeResult`
+* Add range aggregation parsing and hydration in `EntityAggregator`
+* Add range aggregation handling in AggregationParser
+* Add test on the aggregation range using database
+* Add test on the range aggregation in the AggregationParser
+___
+# Elasticsearch
```
Comment (shyim) [inline changelog/_unreleased/2022-09-02-add-range-aggregation.md]: We don't have an Elasticsearch section. Add them to normal Core

File: phpstan-baseline.neon [hunk 2]
```diff
@@ -19250,6 +19250,11 @@ parameters:
 			count: 1
 			path: src/Core/Framework/DataAbstractionLayer/Search/Parser/AggregationParser.php
 
+		-
+			message: "#^Parameter \\#2 \\$field of class Shopware\\\\Core\\\\Framework\\\\DataAbstractionLayer\\\\Search\\\\Aggregation\\\\Metric\\\\RangeAggregation constructor expects string, string\\|null given\\.$#"
```
Comment (shyim) [inline phpstan-baseline.neon]: Please don't add new errors. can you solve it?

File: src/Core/Framework/DataAbstractionLayer/Dbal/EntityAggregator.php [hunk 3]
```diff
@@ -318,8 +327,12 @@ private function parseFilterAggregation(FilterAggregation $aggregation, QueryBui
         $this->extendQuery($aggregation->getAggregation(), $query, $definition, $context);
     }
 
-    private function parseDateHistogramAggregation(DateHistogramAggregation $aggregation, QueryBuilder $query, EntityDefinition $definition, Context $context): void
-    {
+    private function parseDateHistogramAggregation(
```
Comment (shyim) [inline src/Core/Framework/DataAbstractionLayer/Dbal/EntityAggregator.php]: Could you revert the style changes? So it's easier to see your differences?

File: src/Elasticsearch/Test/Product/ElasticsearchProductTest.php [hunk 4]
```diff
@@ -3106,6 +3143,50 @@ public function variantListingConfigProvider(): iterable
...
+
+        $rangesExpectedResult = [
+            '*-*' => 7,
+            'all' => 7,
+            'custom_key' => 3,
+            '*-100' => 1,
+            '100-160' => 2,
+            '200-500' => 4,
+            '*-500' => 7,
+        ];
```
Comment (shyim) [inline src/Elasticsearch/Test/Product/ElasticsearchProductTest.php]: That looks like an inlined data provider. Can you add an actual dataProvider like the method above

File: tests/unit/php/Core/Framework/DataAbstractionLayer/Search/Aggregation/Bucket/RangeAggregationTest.php [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    private Context $context;
+
+    protected function setUp(): void
+    {
+        $this->repository = $this->getContainer()->get('product.repository');
+        $this->context = Context::createDefaultContext();
+    }
+
+    public function buildRangeKeyDataProvider(): \Generator
```
Comment (shyim) [inline tests/unit/php/Core/Framework/DataAbstractionLayer/Search/Aggregation/Bucket/RangeAggregationTest.php]:
```suggestion
    public function buildRangeKeyDataProvider(): iterable
```

Comment (shyim) [inline tests/unit/php/Core/Framework/DataAbstractionLayer/Search/Aggregation/Bucket/RangeAggregationTest.php]: You need here a return type phpdoc for phpstan

File: src/Core/Framework/DataAbstractionLayer/Search/Aggregation/Metric/RangeAggregation.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Aggregation\Aggregation;
+
+/**
+ * Allow to aggregate data on predefined ranges values
+ *
+ * Bound are computed as in the elasticsearch range aggregation :
+ *    the "from" will be compared with greater than or equal
+ *    the "to" will be compared with lower than
+ *
+ * @final tag:v6.5.0
```
Comment (shyim) [inline src/Core/Framework/DataAbstractionLayer/Search/Aggregation/Metric/RangeAggregation.php]: Please add directly `final` to new classes

File: src/Core/Framework/DataAbstractionLayer/Search/Aggregation/Metric/RangeAggregation.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+    /**
+     * Add a new range in the aggregation.
+     * If no key is provided, the key will be build using the $from and $to bounds ($this->buildRangeKey())
+     *
+     * @param ?float $from
+     * @param ?float $to
+     * @param ?string $key
+     *
+     * @return $this
```
Comment (shyim) [inline src/Core/Framework/DataAbstractionLayer/Search/Aggregation/Metric/RangeAggregation.php]: You can remove phpdoc which doesn't add more value than the actual method parameter types

File: tests/unit/php/Core/Framework/DataAbstractionLayer/Search/Aggregation/Bucket/RangeAggregationTest.php [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+use Shopware\Core\Framework\Test\TestCaseHelper\ReflectionHelper;
+use Shopware\Core\Framework\Test\TestDataCollection;
+
+/**
+ * @internal
+ * @covers \Shopware\Core\Framework\DataAbstractionLayer\Search\Aggregation\Metric\RangeAggregation
+ */
+class RangeAggregationTest extends TestCase
+{
+    use IntegrationTestBehaviour;
```
Comment (shyim) [inline tests/unit/php/Core/Framework/DataAbstractionLayer/Search/Aggregation/Bucket/RangeAggregationTest.php]: Please move all Tests which needs a service container / connection to the integration folder. Unit folder should work plain without any traits / db connections

File: src/Elasticsearch/Framework/ElasticsearchRangeAggregation.php [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace Shopware\Elasticsearch\Framework;
+
+use ONGR\ElasticsearchDSL\Aggregation\AbstractAggregation;
+use ONGR\ElasticsearchDSL\Aggregation\Type\BucketingTrait;
+
+/**
+ * @internal
+ */
+class ElasticsearchRangeAggregation extends AbstractAggregation
```
Comment (shyim) [inline src/Elasticsearch/Framework/ElasticsearchRangeAggregation.php:13]: Can you add a simple unit test to this that the aggregation is crafted correct?

**Review by shyim — CHANGES_REQUESTED** [review]
Hey,

It looks overall very good 💙 . I added some comments. Thanks!

**Comment by iNem0o** [comment]
Hi @shyim 

i have changed the PR accordingly

**Comment by shyim** [comment]
@iNem0o Did you tested this also with date fields? 🤔

**Comment by iNem0o** [comment]
It will fail because i used float parameters in the `addRange()` method. 
In my opinion this feature need a dedicated workflow with a `DateRangeAggregation` to allow more flexibility using custom date format and Date Math expressions in the ranges definitions as in the Elasticsearch core.

I updated the DAL documentation to precise that the ranges expects numerics.

**Comment by shyim** [comment]
Could you maybe throw an exception when the target field cannot be used for that aggregation?

**Comment by iNem0o** [comment]
Hi @shyim i added the field type validation

**Comment by keulinho** [comment]
Thanks again for your contribution, your PR was merged! :tada: :blue_heart:

**Comment by iNem0o** [comment]
Nice ! :rocket:

Le mar. 6 déc. 2022, 07:58, Jonas Elfering ***@***.***> a
écrit :

**Pull request closed without merging**